lib/pages/perkPage.dart

```diff
@@ -1,55 +1,91 @@
+import 'package:perklight/classes/perkSerialiser.dart';
 
 class PerkPage extends StatefulWidget {
   PerkPage(arguments) :
     killerPerks = arguments['killerPerks'],
     survivorPerks = arguments['survivorPerks'];
 
-  final List<Perk> killerPerks;
-  final List<Perk> survivorPerks;
+  final List<KillerPerk> killerPerks;
+  final List<SurvivorPerk> survivorPerks;
 
   @override
   _PerkPageState createState() => _PerkPageState();
 }
 
 class _PerkPageState extends State<PerkPage> {
+  static const numPerksToSelect = 4;
+  static const ALL_TILES = [0, 1, 2, 3];
+
+  bool perkModeSwitch = false;
+
   List<int> randomlySelectedPerks;
-  int numPerksToSelect = 4;
-  bool killerMode = false;
+  PerkType perkMode = PerkType.survivor;
 
-  void generateRandomlySelectedPerks() {
-    randomlySelectedPerks = Utils.generateSetOfRandomNumbers(numPerksToSelect, max: perkList.length);
-  }
+  String buildId;
 
-  List<Perk> perkList;
+  List<Perk> rollablePerks;
+  List<Perk> selectedPerks;
 
   @override
   void initState() {
     super.initState();
-    setPerkListAndRandomise();
+
+    _filteredRoll();
+  }
+
+  void _filteredRoll([ List<int> indexes = ALL_TILES ]) {
+    _filterRollable();
+    _rollPerks(indexes);
+  }
+
+  void _generateShareCode() {
+    List<int> selectedPerksIds = selectedPerks.map((item) => item.id).toList();
+    buildId = PerksSerialiser.encode(perksIds: selectedPerksIds, perkType: perkMode);
+
+    print('Build ID: $buildId');
   }
 
-  void setPerkListAndRandomise() {
-    List<Perk> listToFilter = killerMode ? widget.killerPerks : widget.survivorPerks;
-    perkList = listToFilter.where((perk) {
+  void _filterRollable() {
+    List<Perk> listToFilter;
+
+    switch (perkMode) {
+      case PerkType.killer:
+        listToFilter = widget.killerPerks;
+        break;
+      case PerkType.survivor:
+        listToFilter = widget.survivorPerks;
+        break;
+    }
+
+    rollablePerks = listToFilter.where((perk) {
       return perk.preference.enabled;
     }).toList();
-    generateRandomlySelectedPerks();
+
+    randomlySelectedPerks = Utils.generateSetOfRandomNumbers(numPerksToSelect, max: rollablePerks.length);
+  }
+
+  void _rollPerks([ List<int> indexes = ALL_TILES ]) {
+    randomlySelectedPerks = Utils.replaceIndexValue(randomlySelectedPerks, indexes, max: rollablePerks.length);
+    selectedPerks = randomlySelectedPerks.map((item) => rollablePerks[item]).toList();
+
+    _generateShareCode();
   }
 
   void _rollTileCallback(int index, BuildContext context) {
-    // Check if randomlySelectedPerks is less than or equal to perkList
-    if (perkList.length <= randomlySelectedPerks.length) {
-      String perkType = killerMode ? 'killer' : 'survivor';
-      String message = 'You cannot reroll with only 4 $perkType perks selected';
+    // Check if randomlySelectedPerks is less than or equal to rollablePerks
+    if (rollablePerks.length <= numPerksToSelect) {
+      String mode = describeEnum(perkMode);
```

Comment:
You have covered the condition of a user trying to roll a single tile when there are only 4 perks selected. However, you can now press the randomise button and nothing will happen ( i.e. no snackbar)  could the logic for this be added? )

Reply:
Applied